### PR TITLE
Fix DPO and KTO Liger loss gaps

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Fixes DPO and KTO Liger loss gaps found while reviewing #7059:

- forwards DPO label smoothing and DiscoPOP tau;
- fixes KTO metrics for multi-example chunks;
- raises for non-default KTO class weights instead of silently ignoring them;
- keeps the KTO reference path enabled for PEFT, where `ref_model=None` means the disabled adapter is the reference (#5876).

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

We welcome the use of AI tools to help with contributions. For transparency and to help us improve our review process, please indicate the level of AI involvement in this PR.

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Changes training loss and logged metrics for Liger DPO/KTO (including PEFT reference handling); incorrect behavior was fixed, but runs may differ from before.
> 
> **Overview**
> Closes gaps in **Liger fused losses** for DPO and KTO so training behavior matches the non-Liger paths and chunked batches report metrics correctly.
> 
> **DPO:** `DPOTrainer` now passes `label_smoothing` and `discopop_tau` into `FusedLinearDPOLoss` when `use_liger_kernel=True` (they were previously ignored).
> 
> **KTO:** Non-default `desirable_weight` / `undesirable_weight` with the Liger kernel now **fail at init** with a clear error instead of being silently dropped. Liger KTO is always built with `use_ref_model=True` so PEFT setups where `ref_model=None` still use the disabled-adapter reference path. In fused unpaired/KTO code, chosen/rejected **log-prob, logit, and reward sums** fix label broadcasting when `log_probs` is per-sequence and `chunk_size` > 1.
> 
> Tests cover DPO Liger config forwarding, KTO Liger + PEFT `use_ref_model`, weighted KTO + Liger errors, and smaller `chunk_size` in KTO loss correctness tests.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f9d7e85d0c17e2738859dc05bc50c9390328ed62. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
Thanks!

The KTO/DPO half (30c89da5, 72c59fc5) is exactly what the 4 open threads on #7059 need, let's merge that right after #7059.

Can we drop the JSD projection commits from this one? Two real issues: 
- torch.mm is 2-D only and GKD passes (B, T, H) and
- the backward casts grad_output to bf16 before the reductions, so grads don't get the fp32 benefit

If you want it in the meantime, (x.float() @ w.float().t()) like the chunked path is enough I think, no custom Function.

File: trl/losses/fused_linear_distillation.py [hunk 1]
```diff
@@ -24,9 +24,43 @@
...
+        return torch.mm(inputs, weight.t(), out_dtype=torch.float32)
+
+    @staticmethod
+    def setup_context(ctx, inputs, output):
+        ctx.save_for_backward(*inputs)
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        inputs, weight = ctx.saved_tensors
+        grad_output = grad_output.to(inputs.dtype)
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/losses/fused_linear_distillation.py]: Fixed in 3e93baaf. The backward matmuls now run in float32 and only cast the completed gradients back. I also added a regression with small upstream gradients.

File: trl/losses/fused_linear_distillation.py [hunk 2]
```diff
@@ -24,9 +24,43 @@
...
+        return grad_output @ weight, grad_output.t() @ inputs
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/losses/fused_linear_distillation.py]: Fixed in 3e93baaf. The accelerated path now flattens leading dimensions for the 2D matmul and restores them afterward. The test covers 3D inputs, and the fused GKD training tests pass.

**Comment by kashif** [comment]
Done, I dropped the JSD commits. This is now limited to the DPO/KTO fixes; the 48 KTO loss cases and 4 focused trainer tests pass.

**Comment by qgallouedec** [comment]
lgtm, thanks for splitting out the JSD half. 

2 things for the description before merge:
- the class weights are a raise, not support, so say so explicitly, it's a new hard error for anyone running fused KTO with non-default weights today. 
- worth noting the PEFT bug existed because `ref_model is None` meant both "no reference" and "the reference is the disabled adapter", (which is what #5876 is about)

Merging right after #7059.

**Comment by kashif** [comment]
Rebased onto the latest `vendor-chunked-loss`, resolved the `use_liger_kernel` rename conflicts, and updated the description with both points. The 52 focused tests pass.

**Comment by qgallouedec** [comment]
LGTM, thanks!

**Review by qgallouedec — APPROVED** [review]

**Review by qgallouedec — APPROVED** [review]
LGTM, thanks.
